pkg/ccl/backupccl/restore_job.go

```diff
@@ -2010,51 +2018,98 @@ func insertStats(
-				}
+	startingStatsInsertion := timeutil.Now()
+	// Should never block.
+	batchCh := make(chan []*stats.TableStatisticProto, totalNumBatches)
+	generateStatsBatch := func(ctx context.Context) error {
+		defer close(batchCh)
+		for {
+			if len(latestStats) == 0 {
 				return nil
 			}
+			if len(latestStats) < batchSize {
+				batchSize = len(latestStats)
+			}
+			batchCh <- latestStats[:batchSize]
 
-			return nil
-		}); err != nil {
-			return err
+			// Truncate the stats that we have inserted in the txn above.
+			latestStats = latestStats[batchSize:]
 		}
+	}
 
-		// Truncate the stats that we have inserted in the txn above.
-		latestStats = latestStats[restoreStatsInsertBatchSize:]
-
-		if insertStatsProgress.ShouldLog() {
-			remainingBatches := len(latestStats) / restoreStatsInsertBatchSize
-			log.Infof(ctx, "restore has %d/%d TableStatistics batches remaining to insert", remainingBatches, totalNumBatches)
+	logStatsProgress := func(remainingBatches, completedBatches int) {
+		msg := fmt.Sprintf("restore has %d/%d TableStatistics batches remaining to insert",
+			remainingBatches, totalNumBatches)
+		var rate int
+		timeSinceStart := int(timeutil.Since(startingStatsInsertion).Seconds())
+		if completedBatches != 0 && timeSinceStart != 0 {
+			rate = completedBatches / timeSinceStart
+			msg = fmt.Sprintf("%s; ingesting at the rate of %d batches/sec", msg, rate)
 		}
+		log.Infof(ctx, "%s", msg)
+	}
+
+	ingestStatsBatch := func(ctx context.Context) error {
+		mu := struct {
+			syncutil.Mutex
+			completedBatches int
+		}{}
+		return ctxgroup.GroupWorkers(ctx, int(restoreStatsInsertionConcurrency.Get(&execCfg.Settings.SV)),
+			func(ctx context.Context, i int) error {
+				for {
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case b, ok := <-batchCh:
+						if !ok {
+							return nil
+						}
+						if err := execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+							if err := stats.InsertNewStats(
+								ctx, execCfg.Settings, txn, b,
+							); err != nil {
+								return errors.Wrapf(err, "inserting stats from backup")
+							}
+							return nil
+						}); err != nil {
+							return err
+						}
+						mu.Lock()
+						mu.completedBatches++
+						remainingBatches := totalNumBatches - mu.completedBatches
+						completedBatches := mu.completedBatches
+						mu.Unlock()
+						if insertStatsProgress.ShouldLog() {
+							logStatsProgress(remainingBatches, completedBatches)
+						}
+					}
+				}
+			})
+	}
+	if err := ctxgroup.GoAndWait(ctx, generateStatsBatch, ingestStatsBatch); err != nil {
+		return err
 	}
+	logStatsProgress(0, totalNumBatches)
+
+	// Mark the stats insertion complete.
+	return execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		details.StatsInserted = true
+		if err := job.WithTxn(txn).SetDetails(ctx, details); err != nil {
+			return errors.Wrapf(err, "updating job marking stats insertion complete")
+		}
+		return nil
+	})
```

Comment:
[nit] If you want, you can use `job.NoTxn().SetDetails which will create the txn for you. It doesn't matter much.